consensus/src/creation/creator.rs

```diff
@@ -4,85 +4,109 @@ use crate::{
 };
 use log::trace;
 
+#[derive(Clone)]
+struct UnitsCollector<H: Hasher> {
+    n_candidates: NodeCount,
+    candidates: NodeMap<H::Hash>,
+}
+
+impl<H: Hasher> UnitsCollector<H> {
+    pub fn new(n_members: NodeCount) -> Self {
+        Self {
+            n_candidates: NodeCount(0),
+            candidates: NodeMap::with_size(n_members),
+        }
+    }
+
+    pub fn add_unit(&mut self, unit: &Unit<H>) {
+        let pid = unit.creator();
+        let hash = unit.hash();
+
+        if self.candidates.get(pid).is_none() {
+            self.candidates.insert(pid, hash);
```

Comment:
not a comment to this pr but I feel like we should check whether the pid is in range of candidates size (or change the signature of nodemap to return bool whether insert worked)